### ISSUE TITLE
Regularized Greedy Forest is not in the lightgbm  

### ISSUE THREAD

**Issue by anddelu** [microsoft/LightGBM#315]
I heard that Regularized Greedy Forest (RGF) have been used to win some kaggle competitions and it works better than gradient boosting  in some situations . Apparently, up to now, rgf is not included in lightgbm. I hope lightgbm can support it as soon as possible.

**Comment by gugatr0n1c** [comment]
Voting for this as well

**Comment by avannaldas** [comment]
Hi, I can take this up. Will read about RGFs and get back in couple of days.

**Comment by avannaldas** [comment]
Hi @guolinke  , 

I went through the paper [Learning Nonlinear Functions Using Regularized Greedy Forest - arXiv](https://arxiv.org/pdf/1109.0887) which originally proposed RGF. 

The authors of the paper propose three improvements to a Gradient Boosting technique...
1. Introduce regularization, to counter overfitting effects due (2) 
2. Update coefficients/weights repeatedly at every step rather than waiting for an individual tree to build completely. A more aggressive learning
3. Perform greedy search over the forest nodes based on forest structure

Though I've understood RGF in a broad sense, I think I'll have to spend some more time on the paper to understand the intricacies accurately.

Also, the original C++ implementation of RGF is available as a compressed archive [here](http://tongzhang-ml.org/software.html) and uploaded by a GitHub user under a separate folder in [this repo](https://github.com/TimSalimans/HiggsML). I haven't yet looked into this implementation.

I've cloned the LightGBM repo and opened in VS2015, I get an upgrade prompt, which looks optional, hence I skipped upgrade. I'm unable to build the project due to older C++ runtime. Is there a way to build the project in VS2015? Shall I upgrade the project and proceed for now?
![image](https://cloud.githubusercontent.com/assets/11676976/24675878/5efb05fa-199e-11e7-96ac-0bc625d0194a.png)

Also, can you please help me understand LightGBM and share your inputs/guidance on how do we take this forward.

Thanks,
Abhijit Annaldas

**Comment by guolinke** [comment]
Thanks @avannaldas !
you can upgrade to the vs2015 in your local project, but don't push the update of the ```.sln``` and ```.vcproj``` to the online.

for the development guide, you can check the https://github.com/Microsoft/LightGBM/blob/master/docs/development.md . 

It seems the GRF is the improvement of Gradient Boosting. So I think you can just inherit the ```gbdt``` class(https://github.com/Microsoft/LightGBM/blob/master/src/boosting/gbdt.h).

Let me know if you have more questions.

**Comment by avannaldas** [comment]
Thanks for the info @guolinke, I'll spend some more time understanding RGF in depth and go through the existing `gbdt` implementation in LightGBM. Will get back in with questions by Monday.

**Comment by avannaldas** [comment]
Hi @guolinke , I've sent you an email to your Microsoft alias. Thanks!

**Comment by Laurae2** [comment]
@avannaldas Any news?

**Comment by avannaldas** [comment]
I spent some time understanding how this can be done. I had multiple conversations with Guolin about this. Guolin thinks that RGF may not give significant performance improvement over existing GBDT implemenations. @guolinke can share more on this.

**Comment by guolinke** [comment]
Thanks @avannaldas .

Actually, LightGBM already contained many ideas in RGF, e.g. the best-first tree growth, newton step, and so on. So it doesn't have much values for us to implement it.  The remaining difference is their many regularization methods. Maybe we can add these regularization methods in LightGBM. 

However, the contribution is still welcome.

**Comment by StrikerRUS** [comment]
For those who comes here from search engines and searches for the working implementation of the Regularized Greedy Forest (RGF), you could use Python wrapper of the original C++ implementations (both vanilla and fast ones) which were written by the authors of the paper.
https://github.com/fukatani/rgf_python

**Comment by StrikerRUS** [comment]
Closed in favor of being in #2302. We decided to keep all feature requests in one place.

Welcome to contribute this feature! Please re-open this issue (or post a comment if you are not a topic starter) if you are actively working on implementing this feature.
